[1단계 - 로또 구현] 찰리(박민영) 미션 제출합니다.

Gomding opened the pull request:
안녕하세요 데이브!
만나서 반가워요! ✋ 
이번 미션부터 데드라인이 적용되어
빠르게 진행하다가 테스트를 놓친게 많은것 같아요
최대한 테스트를 작성하려고 했지만 습관이란게 무섭네요..

저는 이번에 최대한 테스트코드를 작성하려고 노력해봤어요.
그럼에도 아직 테스트코드를 먼저 짜려고 하면 생각이 꼬이고 겁이 생겨서 피하고싶어지네요 😭 
아직 부족한 테스트들이 있다고 생각해서 데이브가 이 부분에 대한 테스트가 없는데? 라고 생각하시면 바로 찔러주세요
숨어있는 매직넘버 하나까지 탈탈 털어주셔도 좋으니 거침없는 쓴소리 부탁드리겠습니다!

그리고 몇가지 질문이 있어요!

### 질문1
* 현재는 Controller에 입력을 위한 OutputView들이 쓰였는데 이부분을 InputView로 넣어주는게 더 좋을까요? controller의 책임은 아닌것 같아서 걱정이 됩니다.
* -> 현재는 입력에대한 출력은 전부 InputView 안으로 넣었습니다.

### 질문2
* 무조건 커스텀 예외를 만드는것과 자바의 기본 예외를 사용하는것 어느쪽이 더 좋은 방법이라고 할 수 있을까요?
* 내가 던지고 싶은 예외의 의미와 자바의 기본 예외들 중에 있는것과 어느정도 의미가 비슷하다면 최대한 기본 예외를 활용하는 쪽이 더 좋지않을까 해서 질문드립니다!
* 현재는 작은 콘솔 게임이라 만드는 커스텀 예외가 몇개 없지만 큰 서비스라면 엄청나게 많은 커스텀 예외들이 만들어졌을때 낭비가 아닌가 생각했어요.
* 예시) 로또 티켓을 만드는데 생성자의 매개변수로 5개의 숫자만 넘어왔다면
*           IllegalArgumentException 을 던진다. 또는 InvalidLottoException 같은 커스텀 예외를 던진다.

### 질문3
* domain안에서 패키지를 나눠봤는데 옳게 패키징한건지, 패키지를 나눈 기준이 이상하거나 패키지의 이름이 어색한지 데이브의 의견이 필요해요 🙏 
* 제가 domain 안에서 나눈 기준은 아래와 같아요
* domain
       + lottos -> 로또 티켓에 관련된 패키지             [ LottoTicket, LottoTickets, LottoBoughtTicket, LottoNumber ]
           └ + winnerlotto -> 당첨과 관련된 패키지     [ LottoWinner, LottoWinnerTicket, LottoWinnerBonusNumber ]
           └ + rank -> 로또 등수와 관련된 패키지        [ LottoRank ] 
       + money -> 돈과 관련된 패키지                        [ Money ]

# 학습로그
# equals와 hashcode - 2
## 내용
* LottoTicket 객체가 똑바로 생성되었는지 확인하는 테스트를 위해 equals와 hashcode를 오버라이딩함
* LottoTicket 객체가 내부의 로또 숫자들이 같다고 같은 객체? 라는 의문이 들었음
* equals와 hashcode를 삭제하고 isInstanceOf를 사용해서 객체 생성 테스트를 진행했음
## 링크
- https://github.com/Gomding/java-lotto/commit/c24df3b42e57201687c1814b0fc9cf3007de0b0e
- 참고한 블로그 { http://wonwoo.ml/index.php/post/593 }

# Enum을 활용한 당첨 결과 계산 - 2
##내용
* 로또 등수를 enum으로 구현
* 로또 등수의 각 요소는 ( 만족해야하는 매칭 수(matches), 당첨금(reward) ) 를 필드로 가지도록 구현
## 링크
- https://github.com/Gomding/java-lotto/commit/48ec8d1881406508b00ad510a8befb5f4f2a5408

# TreeMap을 이용해서 Map을 정렬 - 3
## 내용
* 로또 등수를 정렬하여 출력하기 위해 LottoRank 값을 정렬하는 Comparator 생성
* TreeMap을 사용해서 로또 당첨 결과를 LottoRank를 기준으로 정렬해줌
## 링크
- https://github.com/Gomding/java-lotto/commit/bcd971934d071e74ac626d8919d0a863dcb5bcb7

#Stream API 사용 - 4
## 내용
* for문과 forEach문은 최대한 Stream API 를 써서 구현
* 읽기 좋은 코드가 되는것을 경험

Gomding committed:
(refactor) : 필요없는 파일 삭제

Gomding committed:
(doc) : README 기능 구현목록 작성

Gomding committed:
(feat) : LottoNumber.java 및 테스트도 구현
숫자를 생성한다.
Todo : 숫자 예외 검증

Gomding committed:
(feat) : LottoNumber 숫자는 1~45 사이의 숫자여야한다.

Gomding committed:
(feat) : LottoTicket.java 스켈레톤 구현

Gomding committed:
(feat) : 로또 티켓의 숫자는 6개여야하고 중복이 없어야한다.

Commit:
feat: 로또 티켓 생성시 중복 처리 예외 검사 및 정렬 수행

Commit:
feat: LottoTickets.java 로또 티켓들을 가진 객체를 생성

Commit:
feat: 유저가 구매한 수량만큼 로또 티켓이 생성되었나 확인한다.

Commit:
refactor: LottoTicketsTest를 BeforeEach활용하여 코드 정리

Commit:
doc: 리드미 체크박스 업데이트

Commit:
feat: 당첨 번호 LottoWinnerTicket 클래스 생성

Commit:
fix: .equals 메서드 수정
IntelliJ에서 generate해서 .equals랑 .hashcode를 자동으로 생성하여 사용하였었는데
LottoWinnerBonusNumber를 만들면서, .equals에서 동등에 대한 처리를 못해주는 버그 발견.
클래스를 비교하는 부분 때문에 발생하였던 버그.

Commit:
feat: 로또 보너스 숫자 생성

Commit:
feat: 로또 당첨번호 생성

Commit:
feat: 당첨 클래스 생성(당첨 티켓과 당첨 보너스 번호를 가지고 있는 클래스)

Commit:
doc: 리드미 업데이트

Gomding committed:
(doc) : README 수정

Gomding committed:
refactor : 로또 당첨번호 6개와 보너스 번호간의 중복검사를 LottoWinnerTicket에 메시지를 보내도록 수정

Gomding committed:
refactor : LottoNumber에 1 ~ 45 의 숫자를 캐싱해서 가지도록 구현

Gomding committed:
feat : LottoTicketService.java 랜덤 번호로 로또 티켓을 생성

Gomding committed:
feat : Money.java 로또 구매금액을 가진다.

Gomding committed:
doc : README 기능구현 체크 수정

Gomding committed:
refactor : Money.java는 1000원 미만은 가질 수 없다.

Gomding committed:
style : 쓸모없는 파일 삭제

Gomding committed:
(refactor) : LottoTicket null검사 추가

Gomding committed:
refactor : LottoTickets null 과 empty 값을 생성자의  매개변수로 받을 때 검사 구현

Gomding committed:
refactor : LottoTicket에 빈값을 체크하는 유효성 검사 추가

Gomding committed:
feat : LottoRank.java 로또 등수와 보상을 모아놓은 enum 구현

Gomding committed:
chore : 패키지 구조 변경

Gomding committed:
chore : import 최적화

Gomding committed:
feat : InputView.java 유저 입력 구현

Gomding committed:
feat : OutputView.java 상수 추가

Commit:
doc: 리드미 업데이트

Commit:
doc: 리드미 수정

Commit:
fix: InputView.java 잘못 들어있었던 throw exception 제거

Commit:
feat: 상수 랩핑, 로또 구매 갯수 세어주는 메서드 추가

Commit:
feat: 유저에게 인풋을 받아서 로또 몇개 구매했는지 출력하는부분까지 완료

Commit:
feat: 구매한 로또 출력

Commit:
feat: 당첨 번호 입력 및 보너스숫자 입력 완료, 당첨 통계 보드 출력

Gomding committed:
feat : null과 빈값 예외처리 구현

Gomding committed:
style : spacing 추가 및 메서드 접근제어자 기준으로 순서 변경

Gomding committed:
refactor : equals hashcode 제거 및 객체 생성관련 테스트 수정
isEqualsTo() -> isInstanceOf() 로 변경
sort테스트 로직 수정

Gomding committed:
doc : 리드미 수정

Gomding committed:
feat : LottoResultStatistic.java 당첨 결과를
enum을 활용하여 map에 갱신함

Commit:
feat: treeMap 사용하여서 결과 출력시 정렬되게끔 구현
todo: 2등 출력 메세지 수정 및 try-catch로 exception 핸들링

Commit:
feat: 2등 당첨부분 로직 수정 완료
todo: 수익률 계산 및 리팩토링

Gomding committed:
feat : 로또 당첨 수익률 계산을 구현

Gomding committed:
feat : 로또 총 수익률 출력 구현

Gomding committed:
doc : README 업데이트

Gomding committed:
refactor : controller의 lottoStart() 메서드 작은 단위로 분할

Gomding committed:
refactor : LottoTicketService에 있던 매직넘버 상수로 변경

Gomding committed:
style : 스타일 수정

Gomding committed:
style : 주석 제거

Gomding committed:
refactor : LottoResultStatistics의 정적 팩토리 메서드 이름 더 명확하게 변경

Gomding committed:
refactor : LottoRank의 필드 final로 변경

Gomding committed:
style : 빈 라인 정리

Gomding committed:
refactor : LottoRank의 생성자 매개변수와 메서드 매개변수에 final 추가

Gomding committed:
refactor : LottoResultStatistics에 필드와 메서드 매개변수에 final추가 및 forEach문 Iterable 메서드로 변경

Gomding committed:
chore : 안쓰는 import 제거

Gomding committed:
refactor : LottoResultStatisticsTest에 3등 당첨되는 경우도 테스트 추가

Gomding committed:
refactor : LottoTicket 필드와 메서드 매개변수에 final 추가 및 유효성 검사 메서드를 한곳으로 모아줌

Gomding committed:
refactor : LottoTicket 생성할 때 번호 개수 유효성 검사에서 던지는 예외를 IllegalArgumentException으로 변경

Gomding committed:
refactor : 로또 구매할 때 자동 생성되는 티켓의 타입을 LottoBoughtTicket으로 변경(LottoTicket을 상속받음)

Gomding committed:
refactor : LottoTickets 에서 getter를 불변객체 리턴으로 변경

Gomding committed:
refactor : 로또 당첨 결과들을 Map에 넣어주는 책임을 LottoResultStatistics -> LottoTickets로 이동
LottoResultStatistics -> LottoTickets 로 메시지를 보내도록 변경

Gomding committed:
style : 주석 제거

Gomding committed:
refactor : LottoWinner 클래스의 equals 와 hashcode 오버라이딩 메서드 제거

Gomding committed:
refactor : 매직넘버 상수로 변경

Gomding committed:
style : 필요없는 공백 제거와 필요한 공백 생성

Gomding committed:
refactor : LottoTicketsService에서 LottoTicket들을 만들어 주는 메서드의 for문을 IntStream으로 변경

Gomding committed:
feat : LottoTicketServiceTest 에 당첨번호 생성 테스트 구현

Gomding committed:
refactor : 람다식을 메서드 레퍼런스로 변경

Gomding committed:
refactor : 람다식을 메서드 레퍼런스로 변경 및 의미없는 stream() 제거

Gomding committed:
refactor : 의미없는 지역 변수 제거

Gomding committed:
refactor : LottoTicketService 에서 매직넘버를 상수로 변경

Gomding committed:
refactor : 로또 당첨 확인 분기를 Optional API 를 사용해서 수정

Gomding committed:
refactor : LottoRank의 new Integet() 부분 모두 제거

Gomding committed:
refactor : forEach반복문 Stream API로 변경

Gomding committed:
refactor : 내부에서 사용하는 메서드 접근제어자를 private 로 변경

Gomding committed:
refactor : Money 테스트에 구매금액 기준으로 로또티켓 구매 숫자를 반환하는 테스트 작성

Gomding committed:
refactor : LottoNumber에서 숫자 캐시를 초기화하는 로직 Stream API 사용

Gomding committed:
refactor : 보너스 번호와 당첨번호 간의 중복검사를 LottoWinnerBonusNumber 로 이동

Gomding committed:
refactor : InputView 내부 메서드 각각의 입력을 담당하도록 변경

Gomding committed:
(refactor) : 로또 당첨 결과 출력부분 위치 수정 및 간소화

Gomding committed:
refactor : LottoBoughtTicket 내부에서 로또 번호가 당첨번호와 얼마나 일치하는지 계산하는 로직 변경

Gomding committed:
refactor : 로또 티켓을 생성하는 로직에서 Stream collectingAndThen을 활용

Gomding committed:
refactor : 애매한 상수의 이름 변경 VALID_NUMBER_SIZE -> LOTTO_NUMBER_SIZE

Gomding committed:
refactor : 상수 이름 변경, LottoTickets의 메서드 내부 변경

Gomding committed:
refactor : 로또 결과를 저장하는 타입을 TreeMap 에서 EnumMap으로 변경

Gomding committed:
refactor : this 키워드 빠진곳 추가

Gomding committed:
refactor : 로또 티켓과 당첨 번호를 매칭하는 메서드의 내부의 로직을 분리해서 구현

Gomding committed:
style : 코드 정리

Gomding committed:
refactor : 상수 이름 변경

Gomding committed:
refactor : 패키지 구조 변경

Gomding committed:
refactor : 패키지 구조 재변경

Gomding committed:
refactor : 로또 생성기능 Stream.generate() 사용

Gomding committed:
refactor : 테스트 패키지 구조도 메인과 동일하게 변경

Gomding committed:
feat : LottoBoughtTicketTest 에서 1~6등까지 구매한 로또 티켓과 당첨 번호가 잘 매칭되는지 테스트 작성

Gomding committed:
style : 코드 정리

Gomding committed:
refactor : LottoRank를 반환해주는 matchLottoRank 메서드의 구조를 변경

if 문이 있는 메서드를 따로 호출하는 것 보다 filter를 한번 더 사용해보는 쪽으로 변경

dave915 commented:
>질문1
현재는 Controller에 입력을 위한 OutputView들이 쓰였는데 이부분을 InputView로 넣어주는게 더 좋을까요? controller의 책임은 아닌것 같아서 걱정이 됩니다.
-> 현재는 입력에대한 출력은 전부 InputView 안으로 넣었습니다.

InputView또한 입력을 받기위한 View이기 때문에 지금처럼 내부에서 OutputView를 호출하지 않고 `System.out.println`이 있어도 된다고 생각해요! 

>질문2
무조건 커스텀 예외를 만드는것과 자바의 기본 예외를 사용하는것 어느쪽이 더 좋은 방법이라고 할 수 있을까요?
내가 던지고 싶은 예외의 의미와 자바의 기본 예외들 중에 있는것과 어느정도 의미가 비슷하다면 최대한 기본 예외를 활용하는 쪽이 더 좋지않을까 해서 질문드립니다!
현재는 작은 콘솔 게임이라 만드는 커스텀 예외가 몇개 없지만 큰 서비스라면 엄청나게 많은 커스텀 예외들이 만들어졌을때 낭비가 아닌가 생각했어요.
예시) 로또 티켓을 만드는데 생성자의 매개변수로 5개의 숫자만 넘어왔다면
      IllegalArgumentException 을 던진다. 또는 InvalidLottoException 같은 커스텀 예외를 던진다.

관련해서 아래 링크를 참고해 보면 좋을 것 같아요! 
https://rok93.tistory.com/entry/%EC%95%84%EC%9D%B4%ED%85%9C72-%ED%91%9C%EC%A4%80-%EC%98%88%EC%99%B8%EB%A5%BC-%EC%82%AC%EC%9A%A9%ED%95%98%EB%9D%BC

>질문3
domain안에서 패키지를 나눠봤는데 옳게 패키징한건지, 패키지를 나눈 기준이 이상하거나 패키지의 이름이 어색한지 데이브의 의견이 필요해요 


저는 지금 잘 나누어 주신 것 같아요! :)

dave915 commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,64 @@
+            return initMoney();
+        }
+        return money;
+    }
```
아래처럼 구현해도 될 것 같아요 :) 
```suggestion
    private Money initMoney() {
        try {
            return new Money(InputView.getMoneyInput());
        } catch (NullPointerException | IllegalArgumentException e) {
            OutputView.printErrorMessage(e);
            return initMoney();
        }
    }
```

Gomding replied:
굉장히 부끄러운 부분이네요 😳😳😳

01f96c2

dave915 commented on src/main/java/lotto/domain/lottos/winnerlotto/LottoWinnerTicket.java:
```diff
@@ -0,0 +1,13 @@
+
+    public LottoWinnerTicket(List<LottoNumber> lottoNumbers) {
+        super(lottoNumbers);
+    }
```
지금 상태라면 `LottoWinnerTicket`대신 LottoTicket으로 사용해도 무관해보여요 ㅎㅎ

Gomding replied:
LottoBoughtTicket 과 구분하기 위해서 LottoWinnerTicket을 만들었는데 이 부분은 낭비가 되는 부분인가요?!
변수명으로 객체를 나타내주면 충분할까요?

9318d18

dave915 replied:
저는 변수명으로 충분하다고 생각들어요!

dave915 commented on src/main/java/lotto/domain/lottos/winnerlotto/LottoWinner.java:
```diff
@@ -0,0 +1,39 @@
+        this.lottoWinnerBonusNumber = lottoWinnerBonusNumber;
+    }
+
+    public void validateIfWinnerBonusNumberInWinnerTicket
```
`validateDuplicateNumber` 정도도 괜찮지 않을까요?

Gomding replied:
지금보니 유효성검사 변수명이 너무 길어서 읽기가 힘드네요!

3b7bdd8

dave915 commented on src/main/java/lotto/domain/LottoResultStatistics.java:
```diff
@@ -0,0 +1,49 @@
+    private static Map<LottoRank, Integer> setLottoResult() {
+        Map<LottoRank, Integer> lottoResult = new EnumMap<>(LottoRank.class);
+        Arrays.stream(LottoRank.values())
+                .forEach(value -> lottoResult.put(value, DEFAULT_COUNT));
```
`Collection.foreach`와 `Stream.foreach`에 관해 한번 찾아보면 좋을 것 같아요! 

```suggestion
        Arrays.asList(LottoRank.values())
                .forEach(value -> lottoResult.put(value, DEFAULT_COUNT));
```

Gomding replied:
해당내용에 대해서 학습해보니
여기서는 Stream의 생성이 낭비되고 있음을 알게됐어요.
감사합니다 데이브 🙇‍♂️ 

791806c

추가로 Collection.forEach는 쓰레드 세이프하다는걸 알았고
Stream.forEach는 병렬 프로그래밍에 특화된 반복을 한다는걸 알았어요!

https://github.com/woowacourse/java-lotto/pull/265#issuecomment-782615289

dave915 commented on src/main/java/lotto/service/LottoTicketService.java:
```diff
@@ -0,0 +1,38 @@
+        return new LottoBoughtTicket(new RandomNumberGenerator().generateNumbers());
+    }
+
+    public static LottoWinnerTicket createLottoWinnerTicket(String input) {
```
요 메소드 테스트가 없네요. 🧐

Gomding replied:
앗 해당 테스트는 있었는데 혹시 말씀하시는 테스트가 LottoTicketsService에 대한 테스트 일까요?!

b24fc13

dave915 commented on src/test/java/lotto/domain/lottos/LottoBoughtTicketTest.java:
```diff
@@ -0,0 +1,167 @@
+                        new LottoNumber(5),
+                        new LottoNumber(6)
+                )
+        );
```
이 코드도 반복 되는 것 같아요! 

https://crazythink.github.io/2019/02/13/Fixture/

Gomding replied:
엄청난 반복이 있었네요.. 말씀해주셔서 감사합니다! 👍 

658b113

dave915 commented on src/test/java/lotto/domain/lottos/LottoTicketsTest.java:
```diff
@@ -0,0 +1,56 @@
+    public void nullParameterTest(List<LottoTicket> lottoTickets) {
+        assertThatThrownBy(() -> {
+            new LottoTickets(lottoTickets);
+        }).isInstanceOf(NullPointerException.class).hasMessage("null 값은 허용하지 않습니다.");
```
```suggestion
        }).isInstanceOf(NullPointerException.class).hasMessage(LottoTickets.NULL_ERROR_MESSAGE);
```

Gomding replied:
테스트하려는 클래스의 상수를 활용할 수 있었네요...
저는 재활용 가능한 상수들을 가둬놓고 있었군요 😳

f0bfb5c

dave915 reviewed (changes requested):
찰리 안녕하세요! 리뷰를 맡게된 데이브입니다. 
요구사항에 맞춰 구현 잘 해주셨네요 :) 
질문 주신 것과 피드백 몇개 남겼습니다. 확인 부탁드릴게요! 
궁금한점이 있다면 언제든 편하게 DM주세요!

Gomding committed:
refactor : LottoController의 initMoney 메서드 내부의 불필요한 지역변수 제거 및 return 문 위치 변경

Gomding committed:
refactor : LottoWinnerTicket 클래스를 삭제하고 LottoTicket을 사용하도록 변경

Gomding committed:
refactor : LottoWinner 의 중복검사 메서드명 validateIfWinnerBonusNumberInWinnerTicket -> validateDuplicateNumber

Gomding committed:
refactor : LottoWinnerBonusNumber 이름 LottoBonusNumber 로 변경

Gomding committed:
refactor : 보너스 번호로 문자열을 입력받은 경우 에러메시지를 보여주도록 변경

Gomding committed:
refactor : 내부에서만 쓰는 메서드 접근지시자 private로 변경

Gomding commented:
# Collection.forEach 와 Stream.forEach 의 차이 - 3
## 내용
* Stream.forEach문은 Stream의 종료를 위한 연산
     * 로직을 수행하는 역할은 Stream을 반환하는 중간 연산의 역할
     * forEach내부에 로직이 하나라도 추가되면 동시성 보장이 어렵고 가독성이 떨어짐
     * 반복문이 보이거나 사용해야할 때 무작정 Stream.forEach를 사용하려면 잠시 키보드에서 손을 떼고 생각해볼것
* Collection.forEach도 있기때문에 예를들어 List 구현체에 Stream.forEach를 사용하는것은 Stream 객체는 버려지는 오버헤드가 발생
* Collections.forEach는 내부적으로 synchronized 키워드가 붙어있기 때문에 쓰레드 세이프하다.
     * 만약 Collection.forEach가 루프를 도는 사이에 해당 Collection 내부의 데이터에 수정이 생기면 ConcurrentModificationException발생
* Stream.forEach 은 해당 컬렉션의 spliterator 를 사용하는데 이 메서드는 따로 synchronized 키워드가 없다. 
     * 때문에 다른 쓰레드에 의해 데이터 수정이 발생할 수 있고 무조건 마지막 요소까지 반복문을 실행한다.
     * (수정된 상태로 반복문을 수행한다는 의미)
     * 병렬 프로그래밍에 특화됨 (paralleStream 을 볼것)
* 결론
     *  일반적인 반복의 경우(특히 컬렉션에 단지 forEach 반복문이 필요한경우) Collection.forEach를 사용하는것이 좋다.

## 링크
* https://woowacourse.github.io/javable/post/2020-08-31-java-loop/
* https://dundung.tistory.com/247
* https://woowacourse.github.io/javable/post/2020-05-14-foreach-vs-forloop/

Gomding committed:
refactor : 무분별한 Stream forEach 사용에 대해 반성하며 Collection.forEach로 수정

Gomding committed:
refactor : LottoTicketService의 createLottoWinnerTicket 내부 메서드 수정

parseInt 메서드 따로 빼서 입력값이 숫자 외의 값일 때 예외 메세지를 던져줄 수 있도록 수정
숫자 개수에 대한 유효성 검사를 메서드로 분리

Gomding committed:
feat : LottoTicketsService 에 대한 테스트 구현 (LottoTicketsServiceTest.java)

Gomding committed:
refactor : LottoBoughtTicketTest의 테스트 픽스쳐 추가

Gomding committed:
refactor : LottoTicketServiceTest안에 테스트 추가

Gomding committed:
refactor : LottoTicketTest의 LottoTicketSortTest의 DisplayName 수정

Gomding committed:
feat : RandomNumberGenerator 에 대한 테스트 생성

Gomding committed:
feat : LottoRank 에 대한 테스트 추가

Gomding committed:
refactor : 예외 테스트의 hasMessage에 테스트하는 클래스의 상수를 활용하도록 변경 및 LottoWinner 의 중복검사 테스트 삭제

Gomding committed:
refactor : LottoTicket 안에 있는 상수를 재활용 하기

Gomding committed:
refactor : RandomNumberGenerator에도 LottoTicket의 상수 활용하기

Gomding committed:
refactor : InputView에 입력 받는 메시지를 출력해줌

Gomding committed:
refactor : 에러메시지를 포매팅 문자열로 만들어서 재사용에 유리하게 변경

dave915 reviewed (approved):
찰리 안녕하세요! 
피드백 반영 및 학습로그 작성해주신 부분 확인했습니다! 
이만 머지할게요!
